modules/service/src/main/resources/db/migration/V0882__program_calibration_role.sql

```diff
@@ -0,0 +1,228 @@
+            RAISE EXCEPTION 'Library programs must define a description';
+          END IF;
+          NEW.c_semester        := NULL;
+          NEW.c_semester_index  := NULL;
+          NEW.c_science_subtype := NULL;
+        END;
+
+      WHEN NEW.c_program_type = 'science' THEN
+        BEGIN
+          IF NEW.c_proposal_status <> 'not_submitted' THEN
+            IF NEW.c_semester IS NULL THEN
+              RAISE EXCEPTION 'Submitted science programs must define a semester';
+            ELSEIF NEW.c_science_subtype IS NULL THEN
+              RAISE EXCEPTION 'Submitted science programs must define a science subtype.';
+            ELSEIF (NEW.c_program_type != OLD.c_program_type) OR
+                   (NEW.c_semester IS DISTINCT FROM OLD.c_semester) OR
+                   (OLD.c_proposal_status = 'not_submitted' AND NEW.c_semester_index IS NULL) THEN
+              NEW.c_semester_index := next_semester_index('science', NEW.c_semester, NULL);
+            END IF;
+          END IF;
+          NEW.c_instrument   := NULL;
+          NEW.c_library_desc := NULL;
+        END;
+    END CASE;
+
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Calibration programs' targets are marked with the same calibration type
+CREATE OR REPLACE FUNCTION calibration_targets_on_calibration_programs()
+  RETURNS trigger AS $$
+DECLARE
+    calibration_role e_calibration_role;
+BEGIN
+    -- Fetch the value from the source table
+    SELECT c_calibration_role INTO calibration_role
+    FROM t_program
+    WHERE c_program_id = NEW.c_program_id;
+
+    NEW.c_calibration_role := calibration_role;
+
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER ch_target_calibration_target
+  BEFORE INSERT OR UPDATE ON t_target
+  FOR EACH ROW
+  EXECUTE PROCEDURE calibration_targets_on_calibration_programs();
+
+-- Update the entry point for formatting to handle SYS
+CREATE OR REPLACE FUNCTION format_program_reference(
+  ptype           e_program_type,
+  semester        d_semester,
+  index           int4,
+  proposal_status d_tag,
+  science_subtype e_science_subtype,
+  instrument      d_tag,
+  description     text
+)
+RETURNS text AS $$
+BEGIN
+    RETURN CASE
+      WHEN ptype = 'calibration'   OR
+           ptype = 'commissioning' OR
+           ptype = 'engineering'   OR
+           ptype = 'monitoring'    THEN
+          format_semester_instrument_reference(ptype, semester, index, instrument)
+
+      WHEN ptype = 'example' OR
+           ptype = 'library' THEN
+          format_lib_or_xpl_reference(ptype, instrument, description)
+
+      WHEN ptype = 'system' THEN
+          CONCAT('SYS-', description)
+
+      WHEN ptype = 'science' AND proposal_status = 'accepted' THEN
+          format_science_reference(semester, index, science_subtype)
+
+      ELSE
+          NULL
+    END;
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
+
+-- Forbid changing calibration role
+CREATE OR REPLACE FUNCTION forbid_calibration_role_change()
```

Comment:
Does this mean the role must be set once when the program is created?

Reply:
yes I expect the few calibration programs to be created directly rather than through the API or UI